### PR TITLE
OneAPI tests for K-Means initialization

### PR DESCRIPTION
**Pull request by amgrigoriev**

### CODE REVIEW

File: cpp/oneapi/dal/algo/kmeans_init/test/batch.cpp [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        return this->get_policy().is_gpu() && (is_plus_plus_dense || is_parallel_plus_dense);
+    }
+
+    bool is_dense() {
+        constexpr bool ret = std::is_same_v<Method, method::dense> ||
+                             std::is_same_v<Method, method::random_dense> ||
+                             std::is_same_v<Method, method::plus_plus_dense> ||
+                             std::is_same_v<Method, method::parallel_plus_dense>;
+        return ret;
+    }
```
Comment (makart19) [inline cpp/oneapi/dal/algo/kmeans_init/test/batch.cpp:51]:
```suggestion
    constexpr bool is_dense() const {
        return std::is_same_v<Method, method::dense> ||
                   std::is_same_v<Method, method::random_dense> ||
                   std::is_same_v<Method, method::plus_plus_dense> ||
                   std::is_same_v<Method, method::parallel_plus_dense>;
    }
```

**Review by makart19 — APPROVED** [review]

File: cpp/oneapi/dal/algo/kmeans_init/test/batch.cpp [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        check_results(cluster_count, data, compute_results.get_centroids());
+    }
+
+    void check_results(std::int64_t cluster_count, const table& data, const table& centroids) {
+        std::int64_t row_count = data.get_row_count();
+        std::int64_t column_count = data.get_column_count();
+        ONEDAL_ASSERT(centroids.get_row_count() == cluster_count);
+        ONEDAL_ASSERT(centroids.get_column_count() == column_count);
+        const auto data_ptr = row_accessor<const float>(data).pull().get_data();
+        const auto centroid_ptr = row_accessor<const float>(centroids).pull().get_data();
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/kmeans_init/test/batch.cpp:69]: ❗ This is wrong use of `dal::array`. The problem is the object returned by `.pull()` is temporal. It's destroyed at the same line. As a result you'll get a dangling pointers to `data` and `centroids`.

Tip: you have read-only access to the array objects itself via overloaded `operator[](std::int64) const`

File: cpp/oneapi/dal/algo/kmeans_init/test/batch.cpp [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        const auto desc = get_descriptor(cluster_count);
+
+        INFO("compute");
+        const auto compute_results = compute(desc, data);
+        check_results(cluster_count, data, compute_results.get_centroids());
+    }
+
+    void check_results(std::int64_t cluster_count, const table& data, const table& centroids) {
+        std::int64_t row_count = data.get_row_count();
+        std::int64_t column_count = data.get_column_count();
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/kmeans_init/test/batch.cpp:65]:
```suggestion
        const std::int64_t row_count = data.get_row_count();
        const std::int64_t column_count = data.get_column_count();
```

File: cpp/oneapi/dal/algo/kmeans_init/test/batch.cpp [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+                        match = false;
+                        break;
+                    }
+                }
+                if (match) {
+                    if (std::count(std::begin(indices), std::end(indices), j) == 0) {
+                        indices.push_back(j);
+                        match_count++;
+                        break;
+                    }
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/kmeans_init/test/batch.cpp:86]: I guess, you could use `std::set`, `std::set::insert` and `std::set::count` instead. Or the same, but `std::unordered_set`

**Review by rlnx — CHANGES_REQUESTED** [review]
This is very basic test. Are you going to add more specific tests, e.g., for k-means++, to test specific properties of algorithm?

File: cpp/oneapi/dal/algo/kmeans_init/test/batch.cpp (hunk 2 above)
Comment (amgrigoriev) [reply cpp/oneapi/dal/algo/kmeans_init/test/batch.cpp]: Fixed

File: cpp/oneapi/dal/algo/kmeans_init/test/batch.cpp (hunk 4 above)
Comment (amgrigoriev) [reply cpp/oneapi/dal/algo/kmeans_init/test/batch.cpp]: Done

**Review by rlnx — APPROVED** [review]

**Pull request merged**